DOC: Add documentation on codespaces usage

melissawm opened the pull request:
## PR summary

This PR adds a skeleton for codespaces usage documentation. There are definitely improvements to be made but I have a couple of points:
- I've added a codespaces.md document instead of an .rst one because I feel like it will be possible to preview this markdown file automatically when the codespace is opened (see [this topic](https://github.com/orgs/community/discussions/59435) and [this comment](https://github.com/matplotlib/matplotlib/issues/26169#issuecomment-1607607764) (which is not generally possible with rst files)
- I have included a minimal set of instructions on the main building page without the desktop preview instructions because I feel that would be too much for new contributors, but I am open to any suggestions.


Addresses #26169

## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at https://matplotlib.org/devdocs/devel/development_workflow.html

- Create a separate branch for your changes and open the PR from this branch. Please avoid working on `main`.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  https://matplotlib.org/stable/devel/documenting_mpl.html#formatting-conventions.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

jklymak commented:
The tab on develop looks good.  Why is the information then duplicated on a second page?

melissawm commented:
Because the `codespaces.md` file will hopefully be opened automatically when codespaces is fired up (which we can't do with rst files)

jklymak commented:
Is that a usual thing for codespaces?  Seems strange to pop open an editable file that is part of the repo. 

Regardless id just have the markdown file have a link to the canonical instructions and not repeat info.

tacaswell commented on doc/devel/codespaces.md:
````diff
@@ -0,0 +1,107 @@
+   `mpl-dev`:
+
+   ```
+   conda activate mpl-dev
````
I think there is a vscode setting file we can drop to tell vscode we want to use the environment and it will auto-activate for us.

melissawm replied:
I will investigate but this was not possible until recently because of the way the terminal is fired up when codespaces is first opened. They have been making a lot of updates lately so I'll verify.

tacaswell commented:
I'm less worried about the duplicate text between the tabs.  I think reading a unified top-to-bottom flow is nicer than reading it broken up.

Is there anything we can do about the navigation on the left?  I am not sure what the right order is, but having "Contribute" below the MEP docs and near the bottom seems less than ideal.  Looking at this whole section I think the bigger problem is less the repetition, but more the disjointedness as you read through the sections in the order on the left....

------

> Regardless id just have the markdown file have a link to the canonical instructions and not repeat info.

If the goal is to make it easy for people to get started, I'm in favor of it being as obviously self documenting in-line as possible.

marbled-toast commented on doc/devel/codespaces.md:
````diff
@@ -0,0 +1,107 @@
+  run (in your `mpl-dev` conda environment):
+
+  ```
+  cd doc
````
Just checking... 
did you want " cd ./doc" instead of "cd doc"?

melissawm replied:
Both should work 😄

melissawm commented on doc/devel/contribute.rst:
```diff
@@ -222,39 +222,96 @@ How to contribute
-4. Enter the directory and install the local version of Matplotlib.
-   See :ref:`installing_for_devs` for instructions
+        3. Clone this copy to your local disk::
 
-5. Create a branch to hold your changes::
+             git clone https://github.com/<YOUR GITHUB USERNAME>/matplotlib.git
 
-      git checkout -b my-feature origin/main
+        4. Enter the directory and install the local version of Matplotlib.
+           See :ref:`installing_for_devs` for instructions
 
-   and start making changes. Never work in the ``main`` branch!
+        5. Create a branch to hold your changes::
 
-6. Work on this copy, on your computer, using Git to do the version control.
-   When you're done editing e.g., ``lib/matplotlib/collections.py``, do::
+             git checkout -b my-feature origin/main
 
-      git add lib/matplotlib/collections.py
-      git commit
+           and start making changes. Never work in the ``main`` branch!
 
-   to record your changes in Git, then push them to GitHub with::
+        6. Work on this copy, on your computer, using Git to do the version
+           control. When you're done editing e.g.,
+           ``lib/matplotlib/collections.py``, do::
 
-      git push -u origin my-feature
+             git add lib/matplotlib/collections.py
+             git commit
+
+           to record your changes in Git, then push them to GitHub with::
+
+             git push -u origin my-feature
+
```
Yeah... my thinking is I would really love to have a tiny quickstart as a reference for codespaces, without asking folks to go through the entire dev docs. But maybe that's the wrong approach and I should be linking to sections instead. Let me think about this

melissawm commented on doc/devel/codespaces.md:
```diff
@@ -0,0 +1,107 @@
+2. Fork the [project repository](https://github.com/matplotlib/matplotlib):
+3. Open codespaces on your fork by clicking on the green "Code" button on the
+   GitHub web interface and selecting the "Codespaces" tab. Next, click on "Open
+   codespaces on <your fork name>". You will be able to change branches later,
```
OOps, I meant branch here. Updating

melissawm commented on doc/devel/codespaces.md:
````diff
@@ -0,0 +1,107 @@
+   python -m pip install -e .
+   ```
+
+   (See [Setting up Matplotlib for development](https://matplotlib.org/devdocs/devel/development_setup.html)
````
This was moved, thanks for pointing it out.

melissawm commented on doc/devel/codespaces.md:
````diff
@@ -0,0 +1,107 @@
+  cd doc
+  make html
+  ```
+
````
Added, thanks!

melissawm commented:
Here's a second take.

I made the codespaces.md file simpler to just link out to the actual contributing guide and reorganized the build instructions a little bit. I think I've addressed most comments but happy to have more feedback. Cheers!

melissawm commented on doc/devel/contribute.rst:
```diff
@@ -222,9 +222,13 @@ How to contribute
-then submit a "pull request" (PR).
+then submit a "pull request" (PR). You can do this by cloning a copy of the
+Maplotlib repository to your own computer, or alternatively using
+`GitHub Codespaces <https://docs.github.com/codespaces>`_ (a cloud-based
+in-browser development environment, that comes with the appropriated setup to
+contribute to Matplotlib).
 
```
Good point! Will fix

melissawm commented on doc/devel/contribute.rst:
```diff
@@ -233,35 +237,86 @@ A brief overview is:
+
+                conda activate mpl-dev
+
+
```
Got it! Finally managed with bullets 😄

QuLogic commented:
I don't feel like doing that backport close to 3.8 release.
